implemented HISTOGRAM metric type that supports custom bins

zyxue opened the pull request:
**What this PR does / why we need it**:

To support user provided custom bins for histogram metrics.

**Which issue(s) this PR fixes**:
Fixes #2663

**Special notes for your reviewer**:
After discussion at https://github.com/SeldonIO/seldon-core/pull/2713, we decided to implement a new metric type `HISTOGRAM` instead of updating the existing metric type `TIMER`.

**Does this PR introduce a user-facing change?**:
```release-note
add HISTOGRAM metric type that supports custom bins.
```

seldondev commented:
Hi @zyxue. Thanks for your PR.

I'm waiting for a [SeldonIO](https://github.com/orgs/SeldonIO/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should [join the org](https://github.com/orgs/SeldonIO/people) to skip this step.

Once the patch is verified, the new status will be reflected by the `ok-to-test` label.

I understand the commands that are listed [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

seldondev commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **ryandawsonuk**
You can assign the PR to them by writing `/assign @ryandawsonuk` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["ryandawsonuk"]} -->

zyxue commented:
Not sure what this error  is about?
```dumping object inventory... done
build finished with problems, 44 warnings.
make: *** [Makefile:57: html] Error 1
make: Leaving directory '/__w/seldon-core/seldon-core/doc'
Error: Process completed with exit code 2.
```

RafalSkolasinski commented on .isort.cfg:
```diff
@@ -0,0 +1,8 @@
+[settings]
+# copied from https://black.readthedocs.io/en/stable/compatible_configs.html#configuration
+multi_line_output = 3
+include_trailing_comma = True
+force_grid_wrap = 0
+use_parentheses = True
+ensure_newline_before_comments = True
+line_length = 88
```
Why do we need black config? If this seems to be default value there is maybe no need to add it?

zyxue replied:
please see https://github.com/SeldonIO/seldon-core/pull/2927. I realize there is a better way to configure compatibility between black and isort now. Will remove this after https://github.com/SeldonIO/seldon-core/pull/2927 is merged.

RafalSkolasinski replied:
Can we separate the two and first get in the `bins`? 

For the `isort`/`black` maybe we could start first with issue describing what is the change and what it adds?

zyxue replied:
I thought it would be easier to get the isort/black thing first. The addition of isort is just to sort imports nothing changed to the logic of the code.

After https://github.com/SeldonIO/seldon-core/pull/2927 is merged, I will rebase and update this pull request. so the two are separated

RafalSkolasinski replied:
Ok, fair enough. Will look at #2927 .

zyxue replied:
If you really prefer to get this PR done first. I'd be fine with it too. please let me know.

RafalSkolasinski replied:
No, it's okay. If this is only sorting imports it's fine. Asked a question on #2927 and triggered integration tests there.

RafalSkolasinski commented on python/README.md:
````diff
@@ -10,4 +10,8 @@
+## Compile protobuf
 
+```
+python -m grpc_tools.protoc -I${PWD}/seldon_core/proto --python_out=${PWD}/seldon_core/proto --grpc_python_out=${PWD}/seldon_core/proto --mypy_out=${PWD}/seldon_core/proto ${PWD}/seldon_core/proto/prediction.proto
````
Ideally, we should only reference the existing Makefile target so we can track it only in one place.

RafalSkolasinski commented on python/seldon_core/metrics.py:
```diff
@@ -127,24 +127,54 @@ def update(self, custom_metrics: List[Dict], method: str):
+                    "tags": tags,
+                }
+            elif metrics_type == "HISTOGRAM":
+                bins = metrics.get("bins")
```
The globally defined `BINS` should come as default here.

RafalSkolasinski commented on python/seldon_core/metrics.py:
```diff
@@ -127,24 +127,54 @@ def update(self, custom_metrics: List[Dict], method: str):
+                    "tags": tags,
+                }
+            elif metrics_type == "TIMER":  # prefer to use HISTOGRAM instead
+                bins = BINS
```
Should also be customisable.

RafalSkolasinski commented on python/seldon_core/metrics.py:
```diff
@@ -127,24 +127,54 @@ def update(self, custom_metrics: List[Dict], method: str):
+
                 # Dividing by 1000 because unit is milliseconds
-                worker_data[key] = {
-                    "value": self._update_hist(metrics["value"] / 1000, vals, sumv),
+                bin_val = metrics["value"] / 1000
```
If the only difference between `TIMER` and new alias `HISTOGRAM` is factor of 1000 they should both fall into 
```
elif metrics_type == "HISTOGRAM" or metrics_type == "TIMER"
```
with one small `if` statement at place where value needs to be modified by factor `1000`.

RafalSkolasinski commented on python/seldon_core/proto/prediction.proto:
```diff
@@ -52,12 +52,16 @@ message Metric {
      COUNTER = 0;
      GAUGE = 1;
-     TIMER = 2;
+     TIMER = 2 [deprecated=true];
```
Don't think we should deprecate.

RafalSkolasinski commented on python/seldon_core/proto/prediction.proto:
```diff
@@ -52,12 +52,16 @@ message Metric {
  float value = 3;
  map<string,string> tags = 4;
+
+ // only used when type is HISTOGRAM
```
Should also be used for `TIMER`

RafalSkolasinski reviewed (changes requested):
Hi @zyxue, 

Thanks for the PR, looks quite good! I added a few comments.

Mainly I think we should not really deprecated the `TIMER`.  Agree, it is a legacy name but it kind of been around with us since before SC 1.0 and it is okay for both `TIMER` and `HISTOGRAM` to to exist simultaneously (with TIMER recording values divided by 1000 for ms unit). 

As this supersede https://github.com/SeldonIO/seldon-core/pull/2713 and https://github.com/SeldonIO/seldon-core/pull/2864 I will close them.

seldondev commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **rafalskolasinski**
You can assign the PR to them by writing `/assign @rafalskolasinski` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["rafalskolasinski"]} -->

RafalSkolasinski commented:
/test integration
/test notebooks

seldondev commented:
Thu Feb  4 14:40:37 UTC 2021
The logs for [integration] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2907/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2907 --build=2

seldondev commented:
Thu Feb  4 14:41:02 UTC 2021
The logs for [notebooks] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2907/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2907 --build=1

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | c7ced5dac6443ad4d0d119930c170c1db7809056 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

seldondev commented:
@zyxue: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | c7ced5dac6443ad4d0d119930c170c1db7809056 | [link]() | `/test integration`
notebooks | c7ced5dac6443ad4d0d119930c170c1db7809056 | [link]() | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

RafalSkolasinski commented:
There seems to be a problem with the `prediction_pb2`. 
Thinking about it more, I do not think we need to include `bins` in the `protos`. 

The metrics returned in the protos was only relevant in the legacy system where metrics was gathered and exposed by deprecated now Java-based orchestrator `engine`. 

I would suggest to revert modification to protos.

zyxue commented:
@RafalSkolasinski , I've addressed the comments, please let me know if you have any more comments.

RafalSkolasinski commented:
/test notebooks
/test integration

seldondev commented:
Sat Feb  6 23:10:19 UTC 2021
The logs for [notebooks] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2907/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2907 --build=3

seldondev commented:
Sat Feb  6 23:10:26 UTC 2021
The logs for [integration] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2907/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2907 --build=4

RafalSkolasinski reviewed (changes requested):
Nice one! We're definitely getting close to merge it ;-).

There are only two minor comments I added.
However, what would be good to include is to extend a  bit docs [here](https://docs.seldon.io/projects/seldon-core/en/latest/analytics/analytics.html#custom-metrics) and [here](https://docs.seldon.io/projects/seldon-core/en/latest/python/python_component.html#adding-custom-metrics) and ideally the example
[notebook](https://docs.seldon.io/projects/seldon-core/en/latest/examples/custom_metrics.html) which uses [this model](https://github.com/SeldonIO/seldon-core/blob/ee2c64173331bd9585995c7aff4fad67a23081b3/examples/models/custom_metrics/ModelWithMetrics.py).


BTW, would you like to show your contribution briefly on Thursday community call @zyxue?

zyxue commented:
> Nice one! We're definitely getting close to merge it ;-).
> 
> There are only two minor comments I added.
> However, what would be good to include is to extend a bit docs [here](https://docs.seldon.io/projects/seldon-core/en/latest/analytics/analytics.html#custom-metrics) and [here](https://docs.seldon.io/projects/seldon-core/en/latest/python/python_component.html#adding-custom-metrics) and ideally the example
> [notebook](https://docs.seldon.io/projects/seldon-core/en/latest/examples/custom_metrics.html) which uses [this model](https://github.com/SeldonIO/seldon-core/blob/ee2c64173331bd9585995c7aff4fad67a23081b3/examples/models/custom_metrics/ModelWithMetrics.py).
> 
> BTW, would you like to show your contribution briefly on Thursday community call @zyxue?

I've addressed the comments, updated `analytics.md` and `python_component.md`. 

But to update the notebook `customMetrics.ipynb`, I'd need to rebuild `seldonio/model-with-metrics` image and push it to DockerHub right, I don't think I have the permission to push though

As for the Thursday community, could you share a bit more on what it is about, please?

RafalSkolasinski commented:
Thanks I will take a look. I can push the image, don't worry. You can for testing build it locally and load into kind environment for testing manually (or if you are using some public k8s cluster just push image to your docker hub account and use that name in your test - just remember to edit back to default one) before that. 

>As for the Thursday community, could you share a bit more on what it is about, please?

We run bi-weekly [community calls](https://docs.seldon.io/projects/seldon-core/en/latest/developer/community.html#community-calls). The agenda is [here](https://docs.google.com/document/d/1gAmiJoD-1NHSr0gvDQBhYGR5ryrgWVWRu3nOOxPdKJk). Thought that if you would want to it would be nice to share the addition with the community.

zyxue commented:
it took me a while to rebuild [`seldon-core-s2i-python36`](https://github.com/SeldonIO/seldon-core/blob/master/examples/models/custom_metrics/Makefile#L6) image with the new version of seldon_core. Since it's not published to PyPi yet, my process is a bit convoluted. I wonder what's the typical practice for rebuilding `seldon-core-s2i-python36`?

I used a local minikube cluster to test `customMetrics.ipynb`, and it works. Btw, I used my own namespaces when building/testing the images like

```
zyxue/model-with-metrics                   0.3-rc.1    f9d4d0135686   16 minutes ago      782MB
zyxue/seldon-core-s2i-python37             1.8.1-dev   cb9845f4e1e4   25 minutes ago      776MB
```

so they're recorded in the output of the notebooks too. Please let me know if that's ok. 

For the community call, sure, I can talk briefly about the addition of `HISTOGRAM`.

RafalSkolasinski commented:
Hmm... We do have the builds here https://hub.docker.com/repository/docker/seldonio/seldon-core-s2i-python36.

I think you could just use `seldon-core-s2i-python36:1.7.0-dev` (published 17 hours ago).

Because we also just released 1.6.0 you may need to rebase on the master.


In regards to the question. We usually build the wrapper images using the scripts in `seldon-core/wrappers/s2i/python/build_scripts` directory.

RafalSkolasinski commented:
A bit comment on 

https://github.com/SeldonIO/seldon-core/blob/5305dee713e0d3f0f028a0c7c6a5a0abdd5dc9b5/examples/models/custom_metrics/Makefile#L6


We use here a tool called [s2i](https://github.com/openshift/source-to-image) which uses `seldonio/seldon-core-s2i-python36`  image as a base to build the `seldonio/model-with-metrics` image.

RafalSkolasinski commented on doc/source/analytics/analytics.md:
```diff
@@ -83,6 +83,12 @@ For this purpose you can supply extra fields in the returned meta data of the re
 				"value": 22.0
 			},
+			{
+				"type": "HISTOGRAM",
```
Would be nice to add one item to list below too.

zyxue replied:
Could you clarify a bit more what you mean?

RafalSkolasinski replied:
A bullet point entry for HISTOGRAM here
```
We provide three types of metric that can be returned in the meta.metrics list:

COUNTER : a monotonically increasing value. It will be added to any existing value from the metric key.
GAUGE : an absolute value showing a level, it will overwrite any existing value.
TIMER : a time value (in msecs), it will be aggregated into Prometheus' HISTOGRAM.
```

RafalSkolasinski replied:
Oh, and it seems there is also a proto snippet just next paragraph.

zyxue replied:
I updated the doc in it

RafalSkolasinski commented on examples/models/custom_metrics/customMetrics.ipynb:
```diff
@@ -35,26 +35,42 @@
    "cell_type": "code",
```
<p>In the PR please commit it as <code>seldonio/model-with-metrics:0.2</code> . </p><p> </p><p>As I can see you are using <code>minikube</code>.  You can build the image as <code>seldonio/model-with-metrics:0.2</code>  using your minikube's docker so it is available right inside it using by executing first</p><pre class="ql-syntax" spellcheck="false">eval $(minikube docker-env)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/SeldonIO/seldon-core/pull/2907/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

zyxue commented:
> seldon-core-s2i-python37:1.7.0-dev

Does `seldon-core-s2i-python37:1.7.0-dev` contain the source code with the `HISTOGRAM` metric enabled? If not, I'd still need to rebuild it I think.

>eval $(minikube docker-env)
yes, this is what I've been using. minikube can pick up local docker images, but I had some issue with `s2i` picking up the `seldon-core-s2i-python3x` docker images from local (even with specifying the `--pull-policy never`). I'll give it a try again.

RafalSkolasinski commented:
> > seldon-core-s2i-python37:1.7.0-dev
> 
> Does `seldon-core-s2i-python37:1.7.0-dev` contain the source code with the `HISTOGRAM` metric enabled? If not, I'd still need to rebuild it I thought.
> 
> > eval $(minikube docker-env)
> 
> yes, this is what I've been using. minikube can pick up local docker images, but I had some issue with `s2i` picking up the `seldon-core-s2i-python3x` docker images from local (even with specifying the `--pull-policy never`). I'll give it a try again.

Oh yes, you are right. It won't have the code. Then the easiest is to use `make build_local` from inside of `seldon-core/wrappers/s2i/python`.

zyxue commented:
@RafalSkolasinski , `make build_local` works. I did `make build` which wouldn't include the current version of source code.

I've rebased and updated the `customMetrics.ipynb` using images from the seldonio namespace.

RafalSkolasinski commented:
/test notebooks
/test integration

seldondev commented:
Thu Feb 11 10:17:05 UTC 2021
The logs for [notebooks] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2907/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2907 --build=5

seldondev commented:
Thu Feb 11 10:17:12 UTC 2021
The logs for [integration] [6] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2907/6.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2907 --build=6

RafalSkolasinski commented:
/test integration

seldondev commented:
Thu Feb 11 13:10:40 UTC 2021
The logs for [integration] [7] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2907/7.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2907 --build=7

RafalSkolasinski commented:
Hi @zyxue, thanks a lot for applying all of the changes!

I am sorry, I made a typo when I asked to use the `seldonio/model-with-metrics:0.2`, what I meant to write was to use it as `seldonio/model-with-metrics:0.3` (modify the version=0.2 to 0.3 in Makefile). Would you mind doing this last change?

zyxue commented:
@RafalSkolasinski , sure, I've updated it to 0.3

zyxue commented:
oh a sec, made a mistake

RafalSkolasinski commented:
Thank you, I am gonna put build the image and put in docker hub and then trigger the testing pipeline :rocket:

zyxue commented:
> Thank you, I am gonna put build the image and put in docker hub and then trigger the testing pipeline 🚀
which image are you referring to, `seldonio/seldon-core-s2i-python37` or `seldonio/model-with-metrics`?

RafalSkolasinski commented:
the example image used in the notebook demo / test

RafalSkolasinski commented:
/test notebooks
/test integration

seldondev commented:
Fri Feb 12 13:41:55 UTC 2021
The logs for [integration] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2907/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2907 --build=9

seldondev commented:
Fri Feb 12 13:42:06 UTC 2021
The logs for [notebooks] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-2907/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-2907 --build=8

zyxue commented:
@RafalSkolasinski 
>For the community call, sure, I can talk briefly about the addition of HISTOGRAM.
I am Sorry that I missed the call. It's 8 am at PST, and the calendar notification didn't show up on my phone. Hope it goes well.

zyxue commented:
Hey @RafalSkolasinski , I wonder if there is anything else pending before merging this PR?

RafalSkolasinski commented:
Hi @zyxue, apologies.  

We had some troubles with the CI system, it was not running the notebook / integration tests for some reason. 
Now things should be fixed, if you'd be so kind to rebase on master we could trigger them then.

zyxue commented:
Sure, I rebased to master and force pushed.

btw, I see a conflict with master,

```python
build:
<<<<<<< HEAD
	s2i build -E environment . seldonio/seldon-core-s2i-python37-ubi8:${PYTHON_WRAPPER} ${IMAGE_BASE}:${VERSION}
=======
	s2i build -E environment . seldonio/seldon-core-s2i-python37:${PYTHON_WRAPPER} ${IMAGE_BASE}:${VERSION}
>>>>>>> 64bab4520... updated to seldon-core-s2i-python37 in examples/models/custom_metrics/Makefile
```

is `seldonio/seldon-core-s2i-python37-ubi8` expected? what does ubi8 mean?

axsaucedo commented:
@zyxue you should now use the ubi8 images, eg seldonio/seldon-core-s2i-python37-ubi8

seldondev commented:
failed to trigger Pull Request pipeline
* failed to create agent
* failed to calculate in repo config
* failed to load trigger config for repository SeldonIO/seldon-core for ref 036839af23c202a5087044bcc8d85dde5a1a733a
* failed to find any lighthouse configuration files in repo SeldonIO/seldon-core at sha 036839af23c202a5087044bcc8d85dde5a1a733a
* failed to process repo SeldonIO/seldon-core refref 036839af23c202a5087044bcc8d85dde5a1a733a
* failed to list files in directory /var/tmp/gitrepo799923688/.lighthouse
* open /var/tmp/gitrepo799923688/.lighthouse
* no such file or directory

zyxue commented:
ok, had some confusion around this image. now it should be good.

RafalSkolasinski commented:
/test notebooks
/test integration

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | e3fb28304a9dea60b7eed805c349d0d869816d4a | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-7jt27) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | e3fb28304a9dea60b7eed805c349d0d869816d4a | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-8bx5f) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

RafalSkolasinski commented:
/test notebooks

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | e3fb28304a9dea60b7eed805c349d0d869816d4a | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-hgrpv) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

RafalSkolasinski commented:
Failing on one of notebook tests
```
=================================== FAILURES ===================================
______________________ TestNotebooks.test_custom_metrics _______________________

self = <test_notebooks.TestNotebooks object at 0x7f6cdf5f3470>

    def test_custom_metrics(self):
        try:
            create_and_run_script(
                "../../examples/models/custom_metrics", "customMetrics"
            )
        except CalledProcessError as e:
            run(
                f"helm delete seldon-core-analytics --namespace seldon-system",
                shell=True,
                check=False,
            )
>           raise e

test_notebooks.py:138: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_notebooks.py:130: in test_custom_metrics
    "../../examples/models/custom_metrics", "customMetrics"
test_notebooks.py:28: in create_and_run_script
    raise e
test_notebooks.py:21: in create_and_run_script
    encoding="utf-8",
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, timeout = None, check = True
popenargs = ('cd ../../examples/models/custom_metrics && ./customMetrics.py',)
kwargs = {'encoding': 'utf-8', 'shell': True, 'stderr': -1, 'stdout': -1}
process = <subprocess.Popen object at 0x7f6cdf5f3550>
stdout = 'Context "kind-kind" modified.\n\x1b[94mapiVersion\x1b[39;49;00m: machinelearning.seldon.io/v1\n\x1b[94mkind\x1b[39;49...0m\x1b[0;34m\x1b[0m\x1b[0m\n\x1b[1;32m    134\x1b[0m \x1b[0;34m\x1b[0m\x1b[0m\n\n\x1b[0;31mKeyError\x1b[0m: \'meta\'\n'
stderr = 'Error: persistentvolumeclaims "seldon-core-analytics-prometheus-alertmanager" not found\nError from server (AlreadyEx...es SQLite, your history will not be saved\n  warn("IPython History requires SQLite, your history will not be saved")\n'
retcode = 1

    def run(*popenargs, input=None, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        The other arguments are the same as for the Popen constructor.
    
        If universal_newlines=True is passed, the "input" argument must be a
        string and stdout/stderr in the returned object will be strings rather than
        bytes.
        """
        if input is not None:
            if 'stdin' in kwargs:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired:
                process.kill()
                stdout, stderr = process.communicate()
                raise TimeoutExpired(process.args, timeout, output=stdout,
                                     stderr=stderr)
            except:
                process.kill()
                process.wait()
                raise
            retcode = process.poll()
            if check and retcode:
                raise CalledProcessError(retcode, process.args,
>                                        output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command 'cd ../../examples/models/custom_metrics && ./customMetrics.py' returned non-zero exit status 1.

/usr/local/lib/python3.6/subprocess.py:438: CalledProcessError
------------------------------ Captured log call -------------------------------
ERROR    root:test_notebooks.py:25 failed notebook test customMetrics stdout:NAME: seldon-core-analytics
LAST DEPLOYED: Wed Feb 24 20:04:48 2021
NAMESPACE: seldon-system
STATUS: deployed
REVISION: 1
Context "kind-kind" modified.
[94mapiVersion[39;49;00m: machinelearning.seldon.io/v1
[94mkind[39;49;00m: SeldonDeployment
[94mmetadata[39;49;00m:
  [94mname[39;49;00m: seldon-model
[94mspec[39;49;00m:
  [94mname[39;49;00m: test-deployment
  [94mpredictors[39;49;00m:
  - [94mcomponentSpecs[39;49;00m:
    - [94mspec[39;49;00m:
        [94mcontainers[39;49;00m:
        - [94mimage[39;49;00m: seldonio/model-with-metrics:0.3
          [94mname[39;49;00m: classifier
    [94mgraph[39;49;00m:
      [94mchildren[39;49;00m: []
      [94mname[39;49;00m: classifier
      [94mtype[39;49;00m: MODEL
    [94mname[39;49;00m: example
    [94mreplicas[39;49;00m: 1
seldondeployment.machinelearning.seldon.io/seldon-model created
Waiting for deployment "seldon-model-example-0-classifier" rollout to finish: 0 of 1 updated replicas are available...
deployment "seldon-model-example-0-classifier" successfully rolled out
[0;31m---------------------------------------------------------------------------[0m
[0;31mKeyError[0m                                  Traceback (most recent call last)
[0;32m/workspace/source/examples/models/custom_metrics/customMetrics.py[0m in [0;36m<module>[0;34m[0m
[1;32m    130[0m [0;32massert[0m [0mlen[0m[0;34m([0m[0mresponse_curl[0m[0;34m)[0m [0;34m==[0m [0;36m1[0m[0;34m[0m[0;34m[0m[0m
[1;32m    131[0m [0mresponse_curl_json[0m [0;34m=[0m [0mjson[0m[0;34m.[0m[0mloads[0m[0;34m([0m[0mresponse_curl[0m[0;34m[[0m[0;36m0[0m[0;34m][0m[0;34m)[0m[0;34m[0m[0;34m[0m[0m
[0;32m--> 132[0;31m [0;32massert[0m [0mlen[0m[0;34m([0m[0mresponse_curl_json[0m[0;34m[[0m[0;34m"meta"[0m[0;34m][0m[0;34m[[0m[0;34m"metrics"[0m[0;34m][0m[0;34m)[0m [0;34m==[0m [0;36m4[0m[0;34m[0m[0;34m[0m[0m
[0m[1;32m    133[0m [0mresponse_curl_json[0m[0;34m[0m[0;34m[0m[0m
[1;32m    134[0m [0;34m[0m[0m

[0;31mKeyError[0m: 'meta'
, stderr:Error from server (AlreadyExists): namespaces "seldon" already exists
/usr/local/lib/python3.6/site-packages/IPython/core/history.py:226: UserWarning: IPython History requires SQLite, your history will not be saved
  warn("IPython History requires SQLite, your history will not be saved")
------------------------------ Captured log call -------------------------------
ERROR    root:test_notebooks.py:25 failed notebook test customMetrics stdout:Context "kind-kind" modified.
[94mapiVersion[39;49;00m: machinelearning.seldon.io/v1
[94mkind[39;49;00m: SeldonDeployment
[94mmetadata[39;49;00m:
  [94mname[39;49;00m: seldon-model
[94mspec[39;49;00m:
  [94mname[39;49;00m: test-deployment
  [94mpredictors[39;49;00m:
  - [94mcomponentSpecs[39;49;00m:
    - [94mspec[39;49;00m:
        [94mcontainers[39;49;00m:
        - [94mimage[39;49;00m: seldonio/model-with-metrics:0.3
          [94mname[39;49;00m: classifier
    [94mgraph[39;49;00m:
      [94mchildren[39;49;00m: []
      [94mname[39;49;00m: classifier
      [94mtype[39;49;00m: MODEL
    [94mname[39;49;00m: example
    [94mreplicas[39;49;00m: 1
seldondeployment.machinelearning.seldon.io/seldon-model created
Waiting for deployment "seldon-model-example-0-classifier" rollout to finish: 0 of 1 updated replicas are available...
deployment "seldon-model-example-0-classifier" successfully rolled out
[0;31m---------------------------------------------------------------------------[0m
[0;31mKeyError[0m                                  Traceback (most recent call last)
[0;32m/workspace/source/examples/models/custom_metrics/customMetrics.py[0m in [0;36m<module>[0;34m[0m
[1;32m    130[0m [0;32massert[0m [0mlen[0m[0;34m([0m[0mresponse_curl[0m[0;34m)[0m [0;34m==[0m [0;36m1[0m[0;34m[0m[0;34m[0m[0m
[1;32m    131[0m [0mresponse_curl_json[0m [0;34m=[0m [0mjson[0m[0;34m.[0m[0mloads[0m[0;34m([0m[0mresponse_curl[0m[0;34m[[0m[0;36m0[0m[0;34m][0m[0;34m)[0m[0;34m[0m[0;34m[0m[0m
[0;32m--> 132[0;31m [0;32massert[0m [0mlen[0m[0;34m([0m[0mresponse_curl_json[0m[0;34m[[0m[0;34m"meta"[0m[0;34m][0m[0;34m[[0m[0;34m"metrics"[0m[0;34m][0m[0;34m)[0m [0;34m==[0m [0;36m4[0m[0;34m[0m[0;34m[0m[0m
[0m[1;32m    133[0m [0mresponse_curl_json[0m[0;34m[0m[0;34m[0m[0m
[1;32m    134[0m [0;34m[0m[0m

[0;31mKeyError[0m: 'meta'
, stderr:Error: persistentvolumeclaims "seldon-core-analytics-prometheus-alertmanager" not found
Error from server (AlreadyExists): namespaces "seldon" already exists
/usr/local/lib/python3.6/site-packages/IPython/core/history.py:226: UserWarning: IPython History requires SQLite, your history will not be saved
  warn("IPython History requires SQLite, your history will not be saved")
```

zyxue commented:
not sure what
```
, stderr:Error: persistentvolumeclaims "seldon-core-analytics-prometheus-alertmanager" not found
```
is about

RafalSkolasinski commented:
I think that one is unrelated, the failing part I believe is 
```
�[0;31mKeyError�[0m: 'meta'
```

zyxue commented:
I see. Is it possible to strip all the non-relevant characters away, it's kind of hard to read the full error

zyxue commented:
It could be the 

> @zyxue you should now use the ubi8 images, eg seldonio/seldon-core-s2i-python37-ubi8

It could be the image, did you update it (version 0.3, I think)  in Dockerhub to include the changes in this PR?

RafalSkolasinski commented:
/test notebooks

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | e3fb28304a9dea60b7eed805c349d0d869816d4a | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-n6kc8) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

RafalSkolasinski commented:
/test notebooks

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | e3fb28304a9dea60b7eed805c349d0d869816d4a | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-d9trg) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

RafalSkolasinski commented:
/test notebooks

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | e3fb28304a9dea60b7eed805c349d0d869816d4a | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-jmfnk) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

RafalSkolasinski commented:
/test notebooks

RafalSkolasinski commented:
That's really odd now. I fixed the failing test - it was indeed my bad, I didn't push the updated example model to docker hub - but now it does fail on following three:
- TestNotebooks.test_jsondata
- TestNotebooks.test_sklearn_iris
- TestNotebooks.test_custom_metrics_server
tests that it didn't fail on before. 

@zyxue would you be so kind to rebase so we can try if that is the root of issues?

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | e3fb28304a9dea60b7eed805c349d0d869816d4a | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-qvws6) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

zyxue commented:
sure, I rebased.

zyxue commented:
@RafalSkolasinski any update on this, please? Is there an easy way for me to see what those notebook tests error are?

RafalSkolasinski commented:
Hi @zyxue, we've been extensively looking into the reasons to fail over last week. 

The notebook test mentioned above was indeed due to the wrong image uploaded to hub, this is now fixed. 

We are however observing failure on three other notebooks that I was unable to reproduce outside of this PR: and even when testing this PR locally. It seems this is a pure flakiness but  we don't understand why yet it manifests itself only here.

The code they are testing should be unrelated to your change so this is even more puzzling.

RafalSkolasinski commented:
/test notebooks

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 7b7fd098e4a2ac8e189c56baccc5e7e9e6a6b801 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-wxltw) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

cliveseldon commented:
/test notebooks

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 7b7fd098e4a2ac8e189c56baccc5e7e9e6a6b801 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-4jwst) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

seldondev commented:
@zyxue: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
notebooks | 7b7fd098e4a2ac8e189c56baccc5e7e9e6a6b801 | [link](http://dashboard-jx.34.66.17.27.nip.io/namespaces/jx/pipelineruns/notebooks-4jwst) | `/test notebooks`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

axsaucedo commented:
Closing as will be revisited with the latest codebase
